File: TelerikUIExamples/TelerikUIExamples/UIKitDynamicsAnimation.m
//
//  UIKitDynamicsAnimation.m
//  TelerikUIExamples
//
//  Copyright (c) 2013 Telerik. All rights reserved.
//

#import "UIKitDynamicsAnimation.h"
#import <TelerikUI/TelerikUI.h>

@interface UIKitDynamicsAnimation() <TKChartDelegate>
@end

@implementation UIKitDynamicsAnimation
{
    TKChart *_chart;
    UIDynamicAnimator *_animator;
    NSMutableArray *_points;
    TKChartVisualPoint *_selectedPoint;
    CGPoint _originalLocation;
    CGPoint _originalPosition;
}

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        [self addOption:@"Apply Gravity" selector:@selector(applyGravity)];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    _chart = [[TKChart alloc] initWithFrame:[self exampleBounds]];
    _chart.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _chart.allowAnimations = YES;
    _chart.delegate = self;
    [self.view addSubview:_chart];
    
    [self reloadChart];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];

    NSArray *points = [_chart visualPointsForSeries:_chart.series[0]];
    TKChartVisualPoint *point = points[4];
    
    UISnapBehavior *snap = [[UISnapBehavior alloc] initWithItem:point snapToPoint:point.center];
    snap.damping = 0.2f;
    
    UIPushBehavior *push = [[UIPushBehavior alloc] initWithItems:@[point] mode:UIPushBehaviorModeInstantaneous];
    push.pushDirection = CGVectorMake(0.f, -1.f);
    push.magnitude = 0.003;
    
    UIDynamicAnimator *animator = [[UIDynamicAnimator alloc] init];
    [animator addBehavior:snap];
    [animator addBehavior:push];
    
    point.animator = animator;
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (CAAnimation *)chart:(TKChart *)chart animationForSeries:(TKChartSeries *)series withState:(TKChartSeriesRenderState *)state inRect:(CGRect)rect
{
    return nil;
}

- (void)applyGravity
{
    _animator = [[UIDynamicAnimator alloc] initWithReferenceView:_chart.plotView];
    
    NSArray *points = [_chart visualPointsForSeries:_chart.series[0]];
    
    UICollisionBehavior *collision = [[UICollisionBehavior alloc] initWithItems:points];
    collision.translatesReferenceBoundsIntoBoundary = YES;
    
    UIGravityBehavior *gravity = [[UIGravityBehavior alloc] initWithItems:points];
    gravity.gravityDirection = CGVectorMake(0.f, 2.f);
    
    UIDynamicItemBehavior *dynamic = [[UIDynamicItemBehavior alloc] initWithItems:points];
    dynamic.elasticity = 0.5f;
    
    [_animator addBehavior:dynamic];
    [_animator addBehavior:gravity];
    [_animator addBehavior:collision];
}

- (void)reloadChart
{
    _points = [[NSMutableArray alloc] init];
    for (int i = 0; i<10; i++) {
        CGFloat x = i * 10;
        CGFloat y = arc4random() % 100;
        TKChartDataPoint *point = [[TKChartDataPoint alloc] initWithX:@(x) Y:@(y)];
        [_points addObject:point];
    }
    
    TKChartLineSeries *lineSeries = [[TKChartLineSeries alloc] initWithItems:_points];
    CGFloat shapeSize = ([UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPhone) ? 14 : 17;
    lineSeries.selectionMode = TKChartSeriesSelectionModeDataPoint;
    lineSeries.style.pointShape = [[TKPredefinedShape alloc] initWithType:TKShapeTypeRhombus andSize:CGSizeMake(shapeSize, shapeSize)];
    lineSeries.style.shapeMode = TKChartSeriesStyleShapeModeAlwaysShow;
    [_chart addSeries:lineSeries];
    _chart.yAxis.style.labelStyle.textHidden = YES;
    
    [_chart reloadData];
}

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event
{
    [super touchesBegan:touches withEvent:event];
    
    UITouch *touch = [touches anyObject];
    CGPoint touchPoint = [touch locationInView:_chart.plotView];
    TKChartSelectionInfo *hitTestInfo = [_chart hitTestForPoint:touchPoint];
    _selectedPoint = [_chart visualPointForSeries:hitTestInfo.series dataPointIndex:hitTestInfo.dataPointIndex];
    _originalLocation = touchPoint;
    if (_selectedPoint) {
        _selectedPoint.animator = nil;
        _originalPosition = _selectedPoint.center;
    }
}

-(void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event
{
    [super touchesMoved:touches withEvent:event];
    
    if (!_selectedPoint) {
        return;
    }
    
    UITouch *touch = [touches anyObject];
    CGPoint touchPoint = [touch locationInView:_chart.plotView];
    CGPoint delta = CGPointMake(_originalLocation.x - touchPoint.x, _originalLocation.y - touchPoint.y);
    
    _selectedPoint.center = CGPointMake(_originalPosition.x, _originalPosition.y - delta.y);
}

- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event
{
    [super touchesEnded:touches withEvent:event];
    
    if (!_selectedPoint) {
        return;
    }
    
    UITouch *touch = [touches anyObject];
    CGPoint touchPoint = [touch locationInView:_chart.plotView];
    CGPoint delta = CGPointMake(_originalLocation.x, _originalLocation.y - touchPoint.y);
    
    UISnapBehavior *snap = [[UISnapBehavior alloc] initWithItem:_selectedPoint snapToPoint:_originalPosition];
    snap.damping = 0.2f;
    
    UIPushBehavior *push = [[UIPushBehavior alloc] initWithItems:@[_selectedPoint] mode:UIPushBehaviorModeInstantaneous];
    push.pushDirection = CGVectorMake(0.f, delta.y > 0 ? -1.f : -1.f);
    push.magnitude = 0.001;
    
    UIDynamicAnimator *animator = [[UIDynamicAnimator alloc] init];
    [animator addBehavior:snap];
    [animator addBehavior:push];
    
    _selectedPoint.animator = animator;
}

@end

File: TelerikUIExamplesInSwift2.0/TelerikUIExamplesInSwift/UIKitDynamicsAnimation.swift
//
//  UIKitDynamicsAnimation.swift
//  Swift Examples
//
//  Copyright (c) 2014 Telerik. All rights reserved.
//

import Foundation
import QuartzCore

class UIKitDynamicsAnimation: ExampleViewController, TKChartDelegate
{
    let chart = TKChart()
    var animator:UIDynamicAnimator?
    var points = [TKChartDataPoint]()
    var selectedPoint:TKChartVisualPoint?
    var originalLocation = CGPoint.zeroPoint
    var originalPosition = CGPoint.zeroPoint
    var location = CGPoint.zeroPoint
    var originalValues = [CGPoint]()
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: NSBundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)

        self.addOption("Apply Gravity") { self.applyGravity() }
    }

    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        chart.frame = self.exampleBoundsWithInset
        chart.autoresizingMask = UIViewAutoresizing(rawValue:UIViewAutoresizing.FlexibleWidth.rawValue | UIViewAutoresizing.FlexibleHeight.rawValue)
        chart.allowAnimations = true
        chart.delegate = self
        self.view.addSubview(chart)
        
        reloadChart()
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        
        let points = chart.visualPointsForSeries(chart.series()[0] as! TKChartSeries)
        
        for x in points {
            let point = x as! TKChartVisualPoint
            originalValues.append(point.CGPoint())
        }
        
        let point = points[4] as! TKChartVisualPoint
        
        location = point.center
        
        let snap = UISnapBehavior(item: point, snapToPoint: point.center)
        snap.damping = 0.2
        
        let push = UIPushBehavior(items:[point], mode:UIPushBehaviorMode.Instantaneous)
        push.pushDirection = CGVectorMake(0, -1)
        push.magnitude = 0.003
        
        let animator = UIDynamicAnimator()
        animator.addBehavior(snap)
        animator.addBehavior(push)
        
        point.animator = animator
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    //MARK: - Methods
    
    func applyGravity() {
        animator = UIDynamicAnimator(referenceView: chart.plotView())
        let points = chart.visualPointsForSeries(chart.series()[0] as! TKChartSeries) as! [UIDynamicItem]

        var i = 0
        for x in points {
            let point = x as! TKChartVisualPoint
            if point.animator != nil {
                point.animator.removeAllBehaviors()
            }
            point.animator = nil
            point.center = originalValues[i]
            i++
        }
        
        let collision = UICollisionBehavior(items: points)
        collision.translatesReferenceBoundsIntoBoundary = true
        
        let gravity = UIGravityBehavior(items: points)
        gravity.gravityDirection = CGVectorMake(0, 2)
        
        let dynamic = UIDynamicItemBehavior(items: points)
        dynamic.elasticity = 0.5
        
        animator!.addBehavior(dynamic)
        animator!.addBehavior(gravity)
        animator!.addBehavior(collision)
    }
    
    func reloadChart() {
        points = [TKChartDataPoint]()
        for i in 0..<10 {
            let x = CGFloat(i*10)
            let y = CGFloat(Int(arc4random() % 100))
            points.append(TKChartDataPoint(x: x, y: y))
        }
        let lineSeries = TKChartLineSeries(items: points)
        let shapeSize = CGFloat(UIDevice.currentDevice().userInterfaceIdiom == .Phone ? 14 : 17)
        lineSeries.selectionMode = TKChartSeriesSelectionMode.DataPoint
        lineSeries.style.pointShape = TKPredefinedShape(type: TKShapeType.Rhombus, andSize: CGSizeMake(shapeSize, shapeSize))
        lineSeries.style.shapeMode = TKChartSeriesStyleShapeMode.AlwaysShow
        chart.addSeries(lineSeries)
        chart.yAxis.style.labelStyle.textHidden = true
        
        chart.reloadData()
    }
    
    //MARK: - TKChartDelegate
    
    func chart(chart: TKChart!, animationForSeries series: TKChartSeries!, withState state: TKChartSeriesRenderState!, inRect rect: CGRect) -> CAAnimation! {
        return nil
    }
    
    //MARK: - Events
    
    override func touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent?) {
        super.touchesBegan(touches, withEvent: event)

        let touch: UITouch = touches.first as UITouch!
        let touchPoint = touch.locationInView(chart.plotView())
        let hitTestInfo = chart.hitTestForPoint(touchPoint)
        if hitTestInfo != nil {
            selectedPoint = chart.visualPointForSeries(hitTestInfo.series, dataPointIndex: hitTestInfo.dataPointIndex)
            originalLocation = touchPoint
            if let point = selectedPoint  {
                point.animator = nil
                originalPosition = point.center
            }
        }
    }
    
    override func touchesMoved(touches: Set<UITouch>, withEvent event: UIEvent?) {
        super.touchesMoved(touches, withEvent: event)
    
        let touch:UITouch = touches.first as UITouch!
        let touchPoint = touch.locationInView(chart.plotView())
        let delta = CGPointMake((originalPosition.x) - (touchPoint.x), originalLocation.y - touchPoint.y)
        if let point = selectedPoint {
            point.center = CGPointMake(originalPosition.x, originalPosition.y - delta.y)
        }
    }
    
    override func touchesEnded(touches: Set<UITouch>, withEvent event: UIEvent?) {
        super.touchesEnded(touches, withEvent: event)

        if let point = selectedPoint {
            let touch:UITouch = touches.first as UITouch!
            let touchPoint = touch.locationInView(chart.plotView())
            let delta = CGPointMake(originalLocation.x, originalLocation.y - touchPoint.y)
            
            let snap = UISnapBehavior(item: selectedPoint!, snapToPoint: originalPosition)
            snap.damping = 0.2
            
            let push = UIPushBehavior(items: [selectedPoint!], mode: UIPushBehaviorMode.Instantaneous)
            push.pushDirection = CGVectorMake(0, delta.y > 0 ? -1 : -1)
            push.magnitude = 0.001
            
            let animator = UIDynamicAnimator()
            animator.addBehavior(snap)
            animator.addBehavior(push)
            
            point.animator = animator
        }
    }
}
